diff --git a/LF/Basics.lean b/LF/Basics.lean
index 1018ff8..894fcee 100644
--- a/LF/Basics.lean
+++ b/LF/Basics.lean
@@ -4,8 +4,6 @@ import SFLMeta
open Verso.Genre Manual
open SFLMeta

[-set_option pp.fieldNotation false-]

#doc (Manual) "Basics: Functional Programming in Lean" =>
%%%
tag := "Basics"
@@ -13,6 +11,10 @@ htmlSplit := .never
file := "Basics"
%%%

{+```lean -show+}
{+set_option pp.fieldNotation false+}
{+```+}

:::instructors
This file and Induction.lean each take about an hour to
get through in a not-too-rushed fashion (with questions, etc.).
@@ -32,7 +34,7 @@ If you don't have Lean installed yet:
* Install Lean 4 through the VS Code extension.
* Start a new terminal session to pick up environment changes.
* Run `make`.
* Run `make serve`. Navigate to [-"http://localhost:8000/lf/student/html-multi/"-]{+"http://localhost:8000/lf/student/html/"+} to start reading.
* Make a copy of "\_out/lf/student/lean" to start solving as if I were a student.
:::

@@ -352,14 +354,7 @@ example : or MyBool.false MyBool.true  = MyBool.true  := by rfl
example : or MyBool.true  MyBool.true  = MyBool.true  := by rfl
```

[-:::dev "mwhicks"-]
[-TODO: Seems wrong-]{+Lean allows us+} to[-not say anything about this notation here.-]
[-Our rule is to mention simple notations like this, but not `macro_rules`-]
[-etc. Do we actually introduce this later?-]
[-:::-]

[-We can-] define[-new-] symbolic [-notations-]{+notation+} for [-existing-]{+our+} definitions.[-Don't worry for now about how the notation is defined.-]

```lean
local prefix:40 (priority := high) "!" => not
@@ -374,6 +369,18 @@ example :
example : (!MyBool.false) = MyBool.true := by rfl
```

{+The technical details of how these symbolic notations work are not something you need to understand until quite a bit later in your Lean journey.  We'll mark these details -- and similar material later on -- with `THESE DETAILS CAN BE SKIPPED` comments in `.lean` files and collapsed text segments in the HTML presentation. Click on the triangle in the HTML if you want to have a peek, or just move on to the following material, as you like.+}

{+:::details+}
{+Lean has a very flexible notation system. Operators like `||` and `&&`+}
{+are defined with specified precedence and associativity. For example, the `infixl` directive above states that+}
{+`&&` is an infix operator, has precedence 35, and is left-associative, while `||` is also infix and left-associative and has precedence 30. This means that `MyBool.true || MyBool.false && MyBool.false` is parsed as `MyBool.true || (MyBool.false && MyBool.false)`.+}

{+Custom notations are defined using the `notation`, `infixl`,+}
{+`infixr`, `prefix`, and `postfix` commands, some of which we will see+}
{+(again, in skippable sections) later on.+}
{+:::+}

:::slidebreak
:::

@@ -549,8 +556,9 @@ theorem true_and_wrong : ∀ (b : MyBool), (MyBool.true && b) = b := by
    rfl
```

[-Lean complains because the {tactic}`rfl` is not at the same level of indentation as-]{+To see+} the [-`{tactic}intro b`,-]
[-so Lean does not recognize these two tactics as being sequential-]{+error message+} in the [-way they-]{+Lean file,+}
{+change `sf_expect_failure` to `sf_expect_failure?` temporarily.+}
{+You+} should [-be.-]{+see the following message.+}

```leanOutput indent
Tactic `introN` failed: There are no additional binders or `let` bindings in the goal to introduce
@@ -559,6 +567,9 @@ b : MyBool
⊢ (true && b) = b
```

{+Lean complains because the {tactic}`rfl` is not at the same level of indentation as the `{tactic}intro b`,+}
{+so Lean does not recognize these two tactics as being sequential in the way they should be.+}

In general, sequential tactics applied to the same goal must be on subsequent lines at the same
level of indentation or separated on the same line by a `;` like so:

@@ -1155,7 +1166,7 @@ Here, though, because {name}`not` is a function that takes a {name}`Bool` argume

::::

[-::::exercise(rating:=0)-]{+::::exercise (rating := 1)+} (name := "custom_namespace_checks")
Predict the output of each of the statements below.
Do you think their results would change depending on which namespace
the statements appear in? How?
@@ -1378,11 +1389,7 @@ inductive Nat : Type where
With a little Lean magic, we can also arrange that
ordinary numerals such as 0, 1, and 2 will be interpreted as values of our new {name}`Nat` type
whenever this is sensible in context.
The technical details of how this is done are not important for present [-purposes,-]
[-so we won't spend time explaining them here.-]
[-Instead, we'll mark them with `THESE DETAILS CAN BE SKIPPED` comments in `.lean` files and-]
[-hide them in a collapsed text segment in the HTML presentation. Click on the triangle in the HTML if you want to have a look.-]{+purposes.+}

:::details "Library Nat to SFL Nat coercion"
```lean
@@ -1391,6 +1398,7 @@ def ofNat : _root_.Nat → Nat
  | .succ n => .succ (ofNat n)

instance (n : _root_.Nat) : OfNat Nat n := ⟨ofNat n⟩
{+attribute [pp_nodot] Nat.succ+}
```
:::

@@ -1524,10 +1532,6 @@ NatPlayground.Nat.succ (NatPlayground.Nat.succ (NatPlayground.Nat.succ (NatPlayg
```

We can also define infix notation for our {name}`add` functions.
[-:::full-]
[-Don't worry too much about how this is defined; we will return to it-]
[-in more detail later.-]
[-:::-]

```lean
scoped infixl:65 " + " => add
@@ -2098,14 +2102,27 @@ one for each of the four cases of control flow through the function.
::::

```lean
theorem [-zero_zero_beq_true-]{+zero_beq_zero+} : (zero == zero) = true := by rfl
theorem [-zero_succ_beq_false-]{+zero_beq_succ+} (n : Nat) : (zero == (succ n)) = false := by rfl
theorem [-succ_zero_beq_false-]{+succ_beq_zero+} (n : Nat) : ((succ n) == zero) = false := by rfl
theorem [-succ_succ_beq-]{+succ_beq_succ+} (n m : Nat) : ((succ n) == (succ m)) = (n == m) := by rfl

attribute [irreducible] beq
```

{+:::full+}
{+As an aside, we point out that we have been following a naming convention+}
{+for simplification rules which aims to convey their meaning.+}
{+For `add_zero` and `add_succ` notice that `zero` and+}
{+`succ` are after the `add` — this is because they depend on `add`'s _second_ argument+}
{+and do not care about its first.+}
{+In the `beq` rules above, we write `zero_beq_zero` and `zero_beq_succ`+}
{+because the rules apply to both the first and second arguments of `beq`. We put+}
{+`beq` in between the arguments because it usually written in infix.+}
{+There are not strict style conventions for naming theorems like this in Lean, but many+}
{+follow this approach.+}
{+:::+}

## General Proofs about Natural Numbers

:::terse
@@ -2244,9 +2261,10 @@ Sometimes simple calculation and rewriting are not enough...


```lean +error
example (n : Nat) : (succ {+zero ++} n == zero) = false := by
  /-
    We can't rewrite by any lemmas {+here: `add`'s definition matches on its+}
{+    *second* argument, and+} here [-because `n`-]{+that argument+} is [-unknown!-]{+the unknown `n`!+}
  -/
```

@@ -2259,13 +2277,13 @@ We can use `cases` to perform case analysis:
:::

```lean
theorem add_one_neb_zero (n : Nat) : (succ {+zero ++} n == zero) = false := by
  cases n with
  | zero =>
    rewrite [-[succ_zero_beq_false]-]{+[add_zero, succ_beq_zero]+}
    rfl
  | succ n' =>
    rewrite [-[succ_zero_beq_false]-]{+[add_succ, succ_beq_zero]+}
    rfl
```

@@ -2302,15 +2320,15 @@ theorem not_involutive (b : Bool) : (!!b) = b := by
```

::::full
[-You may also notice that in-]{+In+} the[-above-] proof {+above+} we have used some rewrite rules that we didn't
previously prove in this [-file!-]{+file.+} These [-proofs-]{+rules+} come from Lean's standard library, in particular
from the section about booleans. Having access to these already-proved theorems about booleans
instead of needing[-them-] to prove them ourselves is [-a big-]{+an+}  advantage of using Lean's built-in
{name}`Bool` type instead of defining our own.

In [-a few chapters-]{+the {ref "UsingLean"}[UsingLean] chapter+} we will discuss how to search through the standard library
for theorems like these. For now, note [-that-]{+that,+} if you hover over the name of these theorems
in [-VSCode,-]{+VS Code,+} the Lean 4 extension will show you their type, i.e., what the theorem proves.
::::

::::terse
@@ -2387,7 +2405,7 @@ theorem and3_exchange (b c d : Bool) :

As you can see, proofs by cases can become very verbose.
We will introduce some tactics for writing shorter proofs
by case analysis in {+the+} {ref "Tactics"}[Tactics] chapter.

## New Tactics: `rewrite ... at` and {tactic}`exact`

@@ -2395,8 +2413,8 @@ by case analysis in {ref "Tactics"}[Tactics] chapter.
Some new tactics will be useful for the exercises ahead.

The `rewrite ... at` tactic can be used to rewrite in a hypothesis instead of the
goal. For example, if `hp : p` is in the context and we have a rule [-`h-]{+`r+} : p = q`,
then `rewrite [-[hp]-]{+[r]+} at [-h`-]{+hp`+} changes the hypothesis to [-`h-]{+`hp+} : q`.

The {tactic}`exact` tactic closes a goal by providing the exact proof of the goal.  For
example, if `hp : p` is in the context and the goal is `p`, then `exact hp`
@@ -2414,10 +2432,9 @@ Prove the following claim.
Tip: the rewrite rule to simplify `(b || false)` is called {name}`Bool.or_false`.

```lean
theorem or_false_true (b : Bool) [-:-]{+(h:+} (b || false) = [-true →-]{+true) :+}
  b = true := by
  solution!
[-    intro h-]
    rewrite [Bool.or_false] at h
    exact h
```
@@ -2429,57 +2446,17 @@ theorem or_false_true (b : Bool) :
::::exercise (rating := 1) (name := "zero_neb_add_one")
```lean
theorem zero_neb_add_one (n : Nat) :
  (zero == [-succ n)-]{+(succ zero + n))+} = false := by
  solution!
    cases n with
    | zero => rewrite [-[zero_succ_beq_false];-]{+[add_zero, zero_beq_succ];+} rfl
    | succ n' => rewrite [-[zero_succ_beq_false];-]{+[add_succ, zero_beq_succ];+} rfl
```

:::gradeTheorem 1 zero_neb_add_one
:::
::::

[-:::dev "Daniel Sainati (dsainati1)"-]
[-I move that we just cut this section entirely and come back to it when-]
[-we've presented enough of the requisite material that we can actually explain-]
[-:::-]

[-:::dev "Michael Hicks (mwhicks1)" BeforeNextRelease-]
[-I'm going to leave this here for now, but perhaps make a note to-]
[-fix later on — when you've fixed it, come back and delete this, rather than-]
[-delete it now.-]
[-:::-]

[-:::dev "Yipeng Liu (berberman)" BeforeNextRelease-]
[-I feel we could split this section and push the typeclass stuff to `Typeclasses` chapter-]
[-and complex notation syntax definitions to TS/HL.-]
[-:::-]

[-## More on Notation (Optional)-]

[-::::full-]
[-Lean has a very flexible notation system.  Operators like `+` and `*`-]
[-are defined with specified precedence and associativity.  For example,-]
[-`+` has precedence 65 and is left-associative, while `*` has-]
[-precedence 70 and is also left-associative.  This means that `1 + 2 * 3 * 4`-]
[-is parsed as `1 + ((2 * 3) * 4)`.-]

[-You can define custom notation using the `notation`, `infixl`,-]
[-`infixr`, `prefix`, and `postfix` commands.-]

[-Lean handles notation scoping through namespaces and _type classes_.-]
[-The numeric literal `3` can be interpreted as {name}`Nat`, {name}`Int`, {name}`Float`, etc.,-]
[-depending on the expected type, thanks to Lean's {name}`OfNat` type class.-]
[-We will explain type classes in more detail in the {ref "Typeclasses"}[Typeclasses] chapter.-]

[-::::-]

[-::::terse-]
[-Lean has commands like `notation`, `infixl`, `infixr`, `prefix`, and `postfix` for defining new-]
[-notation.-]
[-::::-]

## Structural Recursion (Optional)

::::full
@@ -2510,7 +2487,7 @@ automatically, it is sometimes necessary to provide hints or
write functions in slightly different ways.
::::

::::exercise (rating := 2) (name := "decreasing") {+(optional := true)+}
To get a concrete sense of how termination checking works in Lean,
find a way to write a sensible recursive definition (of a simple
function on numbers, say) that does actually terminate on all inputs,
@@ -2521,8 +2498,9 @@ termination.

```lean +error
def factorial_bad (n : Nat) : Nat :=
  [-if-]{+match+} n [-== 0 then 1-]
[-  else-]{+with+}
{+  | zero => (succ zero)+}
{+  | succ _ =>+} n * factorial_bad (pred n)
```

This fails because Lean can't see that `pred n` is structurally smaller.
@@ -2679,7 +2657,7 @@ GRADE_MANUAL 1: negation_fn_applied_twice
:::
::::

::::exercise (rating := 3) (name := "and_eq_or") {+(optional := true)+}
Prove the following theorem.

```lean
