diff --git a/LF/Basics.lean b/LF/Basics.lean
index 1018ff8..894fcee 100644
--- a/LF/Basics.lean
+++ b/LF/Basics.lean
@@ -1,23 +1,25 @@
 prelude
 import SFLMeta
 
 open Verso.Genre Manual
 open SFLMeta
 
-set_option pp.fieldNotation false
-
 #doc (Manual) "Basics: Functional Programming in Lean" =>
 %%%
 tag := "Basics"
 htmlSplit := .never
 file := "Basics"
 %%%
 
+```lean -show
+set_option pp.fieldNotation false
+```
+
 :::instructors
 This file and Induction.lean each take about an hour to
 get through in a not-too-rushed fashion (with questions, etc.).
 
 (N.b. This estimate may need to be revised now that the chapter has been converted to Lean! Please edit this note to reflect your own experience teaching it.)
 
 You may want to assign both files together as the homework for the
 first week, depending on the level of the class.  Just Basics is
@@ -27,17 +29,17 @@ be people that struggle with some of it.
 PRESENTATION ADVICE: Working with the .lean file directly in VS Code
 is recommended for the first few lectures, so students see exactly
 what's in the source file.
 
 If you don't have Lean installed yet:
 * Install Lean 4 through the VS Code extension.
 * Start a new terminal session to pick up environment changes.
 * Run `make`.
-* Run `make serve`. Navigate to "http://localhost:8000/lf/student/html-multi/" to start reading.
+* Run `make serve`. Navigate to "http://localhost:8000/lf/student/html/" to start reading.
 * Make a copy of "\_out/lf/student/lean" to start solving as if I were a student.
 :::
 
 ::::full
 The _functional style_ of programming is founded on simple
 mathematical intuitions: A program is essentially a concrete
 means for computing a mathematical function, which just maps
 inputs to outputs. Even when programs have side effects, such
@@ -347,38 +349,43 @@ Note the syntax for defining multi-argument functions (`and` and `or`).
 
 ```lean
 example : or MyBool.true  MyBool.false = MyBool.true  := by rfl
 example : or MyBool.false MyBool.false = MyBool.false := by rfl
 example : or MyBool.false MyBool.true  = MyBool.true  := by rfl
 example : or MyBool.true  MyBool.true  = MyBool.true  := by rfl
 ```
 
-:::dev "mwhicks"
-TODO: Seems wrong to not say anything about this notation here.
-Our rule is to mention simple notations like this, but not `macro_rules`
-etc. Do we actually introduce this later?
-:::
-
-We can define new symbolic notations for existing definitions.
-Don't worry for now about how the notation is defined.
+Lean allows us to define symbolic notation for our definitions.
 
 ```lean
 local prefix:40 (priority := high) "!" => not
 local infixl:35 (priority := high) " && " => and
 local infixl:30 (priority := high) " || " => or
 ```
 
 ```lean
 example :
     (MyBool.false || MyBool.false || MyBool.true) = MyBool.true := by rfl
 
 example : (!MyBool.false) = MyBool.true := by rfl
 ```
 
+The technical details of how these symbolic notations work are not something you need to understand until quite a bit later in your Lean journey.  We'll mark these details -- and similar material later on -- with `THESE DETAILS CAN BE SKIPPED` comments in `.lean` files and collapsed text segments in the HTML presentation. Click on the triangle in the HTML if you want to have a peek, or just move on to the following material, as you like.
+
+:::details
+Lean has a very flexible notation system. Operators like `||` and `&&`
+are defined with specified precedence and associativity. For example, the `infixl` directive above states that
+`&&` is an infix operator, has precedence 35, and is left-associative, while `||` is also infix and left-associative and has precedence 30. This means that `MyBool.true || MyBool.false && MyBool.false` is parsed as `MyBool.true || (MyBool.false && MyBool.false)`.
+
+Custom notations are defined using the `notation`, `infixl`,
+`infixr`, `prefix`, and `postfix` commands, some of which we will see
+(again, in skippable sections) later on.
+:::
+
 :::slidebreak
 :::
 
 ::::exercise (rating := 1) (name := "nand")
 The {tactic}`sorry` keyword is a placeholder for an incomplete proof or
 definition.  We use it in exercises to indicate the parts that we're
 leaving for you — i.e., your job is to replace {tactic}`sorry` with real
 definitions and proofs.
@@ -544,26 +551,30 @@ it can change the meaning of the proof, usually causing the proof to break. Supp
 instead written the following:
 
 ```lean +error (name := indent)
 theorem true_and_wrong : ∀ (b : MyBool), (MyBool.true && b) = b := by
   intro b
     rfl
 ```
 
-Lean complains because the {tactic}`rfl` is not at the same level of indentation as the `{tactic}intro b`,
-so Lean does not recognize these two tactics as being sequential in the way they should be.
+To see the error message in the Lean file,
+change `sf_expect_failure` to `sf_expect_failure?` temporarily.
+You should see the following message.
 
 ```leanOutput indent
 Tactic `introN` failed: There are no additional binders or `let` bindings in the goal to introduce
 
 b : MyBool
 ⊢ (true && b) = b
 ```
 
+Lean complains because the {tactic}`rfl` is not at the same level of indentation as the `{tactic}intro b`,
+so Lean does not recognize these two tactics as being sequential in the way they should be.
+
 In general, sequential tactics applied to the same goal must be on subsequent lines at the same
 level of indentation or separated on the same line by a `;` like so:
 
 ```lean
 theorem true_and' : ∀ (b : MyBool), (MyBool.true && b) = b := by
   intro b; rfl
 ```
 ::::
@@ -1150,17 +1161,17 @@ Here, though, because {name}`not` is a function that takes a {name}`Bool` argume
 ```
 
 ```leanOutput bt
 !true : Bool
 ```
 
 ::::
 
-::::exercise(rating:=0) (name := "custom_namespace_checks")
+::::exercise (rating := 1) (name := "custom_namespace_checks")
 Predict the output of each of the statements below.
 Do you think their results would change depending on which namespace
 the statements appear in? How?
 
 ```
 #check .black -- Write your prediction here.
 #check Color.black -- Write your prediction here.
 #check RGB -- Write your prediction here.
@@ -1373,29 +1384,26 @@ For simplicity in proofs, we choose a _unary_ representation of natural numbers.
 inductive Nat : Type where
   | zero
   | succ (n : Nat)
 ```
 
 With a little Lean magic, we can also arrange that
 ordinary numerals such as 0, 1, and 2 will be interpreted as values of our new {name}`Nat` type
 whenever this is sensible in context.
-
-The technical details of how this is done are not important for present purposes,
-so we won't spend time explaining them here.
-Instead, we'll mark them with `THESE DETAILS CAN BE SKIPPED` comments in `.lean` files and
-hide them in a collapsed text segment in the HTML presentation. Click on the triangle in the HTML if you want to have a look.
+The technical details of how this is done are not important for present purposes.
 
 :::details "Library Nat to SFL Nat coercion"
 ```lean
 def ofNat : _root_.Nat → Nat
   | .zero => .zero
   | .succ n => .succ (ofNat n)
 
 instance (n : _root_.Nat) : OfNat Nat n := ⟨ofNat n⟩
+attribute [pp_nodot] Nat.succ
 ```
 :::
 
 :::full
 We'll define some shorthands for numbers, putting them in the `Nat` namespace
 so we don't need to use `.` notation everywhere.
 :::
 
@@ -1519,20 +1527,16 @@ def add (n : Nat) (m : Nat) : Nat :=
 #eval add one two -- succ (succ (succ zero)) -- aka, three!
 ```
 
 ```leanOutput three_1
 NatPlayground.Nat.succ (NatPlayground.Nat.succ (NatPlayground.Nat.succ (NatPlayground.Nat.zero)))
 ```
 
 We can also define infix notation for our {name}`add` functions.
-:::full
-Don't worry too much about how this is defined; we will return to it
-in more detail later.
-:::
 
 ```lean
 scoped infixl:65 " + " => add
 ```
 
 ```lean (name := three_2)
 #eval one + two -- succ (succ (succ zero)) -- aka, three again.
 ```
@@ -2093,24 +2097,37 @@ claim.
 ::::
 
 ::::full
 We can also now define the simplification rules for {name}`beq` with our new notation,
 one for each of the four cases of control flow through the function.
 ::::
 
 ```lean
-theorem zero_zero_beq_true : (zero == zero) = true := by rfl
-theorem zero_succ_beq_false (n : Nat) : (zero == (succ n)) = false := by rfl
-theorem succ_zero_beq_false (n : Nat) : ((succ n) == zero) = false := by rfl
-theorem succ_succ_beq (n m : Nat) : ((succ n) == (succ m)) = (n == m) := by rfl
+theorem zero_beq_zero : (zero == zero) = true := by rfl
+theorem zero_beq_succ (n : Nat) : (zero == (succ n)) = false := by rfl
+theorem succ_beq_zero (n : Nat) : ((succ n) == zero) = false := by rfl
+theorem succ_beq_succ (n m : Nat) : ((succ n) == (succ m)) = (n == m) := by rfl
 
 attribute [irreducible] beq
 ```
 
+:::full
+As an aside, we point out that we have been following a naming convention
+for simplification rules which aims to convey their meaning.
+For `add_zero` and `add_succ` notice that `zero` and
+`succ` are after the `add` — this is because they depend on `add`'s _second_ argument
+and do not care about its first.
+In the `beq` rules above, we write `zero_beq_zero` and `zero_beq_succ`
+because the rules apply to both the first and second arguments of `beq`. We put
+`beq` in between the arguments because it usually written in infix.
+There are not strict style conventions for naming theorems like this in Lean, but many
+follow this approach.
+:::
+
 ## General Proofs about Natural Numbers
 
 :::terse
 A (slightly) more interesting theorem:
 :::
 
 
 ::::full
@@ -2239,38 +2256,39 @@ rewriting: In general, the presence of unknown, hypothetical values
 ::::
 
 :::terse
 Sometimes simple calculation and rewriting are not enough...
 :::
 
 
 ```lean +error
-example (n : Nat) : (succ n == zero) = false := by
+example (n : Nat) : (succ zero + n == zero) = false := by
   /-
-    We can't rewrite by any lemmas here because `n` is unknown!
+    We can't rewrite by any lemmas here: `add`'s definition matches on its
+    *second* argument, and here that argument is the unknown `n`!
   -/
 ```
 
 ::::full
 The tactic that tells Lean to consider separate cases is called {tactic}`cases`.
 ::::
 
 :::terse
 We can use `cases` to perform case analysis:
 :::
 
 ```lean
-theorem add_one_neb_zero (n : Nat) : (succ n == zero) = false := by
+theorem add_one_neb_zero (n : Nat) : (succ zero + n == zero) = false := by
   cases n with
   | zero =>
-    rewrite [succ_zero_beq_false]
+    rewrite [add_zero, succ_beq_zero]
     rfl
   | succ n' =>
-    rewrite [succ_zero_beq_false]
+    rewrite [add_succ, succ_beq_zero]
     rfl
 ```
 
 ::::full
 The {tactic}`cases` tactic generates _two_ subgoals, which we must
 prove, separately, in order to get Lean to accept the theorem.
 The generated subgoals are tagged by the names of the constructors.
 `| zero =>` and `| succ n' =>` select which subgoal to work on next
@@ -2297,25 +2315,25 @@ theorem not_involutive (b : Bool) : (!!b) = b := by
     rewrite [Bool.not_false, Bool.not_true]
     rfl
   | true =>
     rewrite [Bool.not_true, Bool.not_false]
     rfl
 ```
 
 ::::full
-You may also notice that in the above proof we have used some rewrite rules that we didn't
-previously prove in this file! These proofs come from Lean's standard library, in particular
+In the proof above we have used some rewrite rules that we didn't
+previously prove in this file. These rules come from Lean's standard library, in particular
 from the section about booleans. Having access to these already-proved theorems about booleans
-instead of needing them to prove them ourselves is a big advantage of using Lean's built-in
+instead of needing to prove them ourselves is an  advantage of using Lean's built-in
 {name}`Bool` type instead of defining our own.
 
-In a few chapters we will discuss how to search through the standard library
-for theorems like these. For now, note that if you hover over the name of these theorems
-in VSCode, the Lean 4 extension will show you their type, i.e., what the theorem proves.
+In the {ref "UsingLean"}[UsingLean] chapter we will discuss how to search through the standard library
+for theorems like these. For now, note that, if you hover over the name of these theorems
+in VS Code, the Lean 4 extension will show you their type, i.e., what the theorem proves.
 ::::
 
 ::::terse
 Some of the above proofs use standard library lemmas; later on we will discuss how to search for
 those yourself.
 ::::
 
 :::slidebreak
@@ -2382,26 +2400,26 @@ theorem and3_exchange (b c d : Bool) :
         rfl
       | true =>
         rewrite [Bool.and_false, Bool.and_true, Bool.and_self]
         rfl
 ```
 
 As you can see, proofs by cases can become very verbose.
 We will introduce some tactics for writing shorter proofs
-by case analysis in {ref "Tactics"}[Tactics] chapter.
+by case analysis in the {ref "Tactics"}[Tactics] chapter.
 
 ## New Tactics: `rewrite ... at` and {tactic}`exact`
 
 ::::full
 Some new tactics will be useful for the exercises ahead.
 
 The `rewrite ... at` tactic can be used to rewrite in a hypothesis instead of the
-goal. For example, if `hp : p` is in the context and we have a rule `h : p = q`,
-then `rewrite [hp] at h` changes the hypothesis to `h : q`.
+goal. For example, if `hp : p` is in the context and we have a rule `r : p = q`,
+then `rewrite [r] at hp` changes the hypothesis to `hp : q`.
 
 The {tactic}`exact` tactic closes a goal by providing the exact proof of the goal.  For
 example, if `hp : p` is in the context and the goal is `p`, then `exact hp`
 closes the goal. You can also transform `hp` slightly when using `exact`, and we will
 explain how when we get to an example that needs it.
 ::::
 
 ::::terse
@@ -2409,82 +2427,41 @@ You will need the `rewrite ... at` and {tactic}`exact` tactics to complete the f
 ::::
 
 ::::exercise (rating := 2) (name := "or_false_true")
 Prove the following claim.
 
 Tip: the rewrite rule to simplify `(b || false)` is called {name}`Bool.or_false`.
 
 ```lean
-theorem or_false_true (b : Bool) :
-    (b || false) = true → b = true := by
+theorem or_false_true (b : Bool) (h: (b || false) = true) :
+  b = true := by
   solution!
-    intro h
     rewrite [Bool.or_false] at h
     exact h
 ```
 
 :::gradeTheorem 2 or_false_true
 :::
 ::::
 
 ::::exercise (rating := 1) (name := "zero_neb_add_one")
 ```lean
 theorem zero_neb_add_one (n : Nat) :
-  (zero == succ n) = false := by
+  (zero == (succ zero + n)) = false := by
   solution!
     cases n with
-    | zero => rewrite [zero_succ_beq_false]; rfl
-    | succ n' => rewrite [zero_succ_beq_false]; rfl
+    | zero => rewrite [add_zero, zero_beq_succ]; rfl
+    | succ n' => rewrite [add_succ, zero_beq_succ]; rfl
 ```
 
 :::gradeTheorem 1 zero_neb_add_one
 :::
 ::::
 
-:::dev "Daniel Sainati (dsainati1)"
-I move that we just cut this section entirely and come back to it when
-we've presented enough of the requisite material that we can actually explain
-:::
-
-:::dev "Michael Hicks (mwhicks1)" BeforeNextRelease
-I'm going to leave this here for now, but perhaps make a note to
-fix later on — when you've fixed it, come back and delete this, rather than
-delete it now.
-:::
-
-:::dev "Yipeng Liu (berberman)" BeforeNextRelease
-I feel we could split this section and push the typeclass stuff to `Typeclasses` chapter
-and complex notation syntax definitions to TS/HL.
-:::
-
-## More on Notation (Optional)
-
-::::full
-Lean has a very flexible notation system.  Operators like `+` and `*`
-are defined with specified precedence and associativity.  For example,
-`+` has precedence 65 and is left-associative, while `*` has
-precedence 70 and is also left-associative.  This means that `1 + 2 * 3 * 4`
-is parsed as `1 + ((2 * 3) * 4)`.
-
-You can define custom notation using the `notation`, `infixl`,
-`infixr`, `prefix`, and `postfix` commands.
-
-Lean handles notation scoping through namespaces and _type classes_.
-The numeric literal `3` can be interpreted as {name}`Nat`, {name}`Int`, {name}`Float`, etc.,
-depending on the expected type, thanks to Lean's {name}`OfNat` type class.
-We will explain type classes in more detail in the {ref "Typeclasses"}[Typeclasses] chapter.
-
-::::
-
-::::terse
-Lean has commands like `notation`, `infixl`, `infixr`, `prefix`, and `postfix` for defining new
-notation.
-::::
-
 ## Structural Recursion (Optional)
 
 ::::full
 Here is a copy of the definition of `even`:
 
 ```lean
 def even' (n : Nat) : Bool :=
   match n with
@@ -2505,29 +2482,30 @@ termination checker, and so this recursive definition is accepted.
 This requirement is a fundamental feature of Lean's design: In
 particular, it guarantees that every function that can be defined
 in Lean will terminate on all inputs.  However, because Lean's
 termination analysis is not always able to figure things out
 automatically, it is sometimes necessary to provide hints or
 write functions in slightly different ways.
 ::::
 
-::::exercise (rating := 2) (name := "decreasing")
+::::exercise (rating := 2) (name := "decreasing") (optional := true)
 To get a concrete sense of how termination checking works in Lean,
 find a way to write a sensible recursive definition (of a simple
 function on numbers, say) that does actually terminate on all inputs,
 but that Lean will reject because it cannot automatically prove
 termination.
 
 :::solution
 
 ```lean +error
 def factorial_bad (n : Nat) : Nat :=
-  if n == 0 then 1
-  else n * factorial_bad (pred n)
+  match n with
+  | zero => (succ zero)
+  | succ _ => n * factorial_bad (pred n)
 ```
 
 This fails because Lean can't see that `pred n` is structurally smaller.
 
 :::
 ::::
 
 ## Binary Numerals
@@ -2674,17 +2652,17 @@ theorem negation_fn_applied_twice (f : Bool → Bool) :
 
 :::grade
 ```
 GRADE_MANUAL 1: negation_fn_applied_twice
 ```
 :::
 ::::
 
-::::exercise (rating := 3) (name := "and_eq_or")
+::::exercise (rating := 3) (name := "and_eq_or") (optional := true)
 Prove the following theorem.
 
 ```lean
 theorem and_eq_or (b c : Bool) : (b && c) = (b || c) → b = c := by
   solution!
     intro h
     cases c with
     | true =>
