--- LF/Lists.lean (before this round)
+++ LF/Lists.lean
@@ -22,7 +22,6 @@
 And UsingLean??
 :::
 
-
 ```importBlock
 import LF.Induction
 import LF.UsingLean
@@ -49,8 +48,8 @@
 
 ::::full
 In an `inductive` type definition, each constructor can take
-any number of arguments — none (as with {name}`true` and  {lean}`0`),
-one (as with  {name}`Nat.succ`), or more than one (as with  {name}`Playground.Nibble` and
+any number of arguments — none (as with {name}`true` and {lean}`0`),
+one (as with {name}`Nat.succ`), or more than one (as with {name}`Playground.Nibble` and
 the following):
 ::::
 
@@ -77,7 +76,7 @@
 :::slidebreak
 :::
 
-:::dev "Mike Hicks (@mwhicks1)"
+:::dev "Mike Hicks (mwhicks1)"
 I would have expected us to have `namespace NatProd` here when defining
 the following functions, so we don't need qualifiers. We've already
 fully explained namespaces back in Basics. Some of the text below
@@ -98,7 +97,7 @@
   | .pair _ y => y
 ```
 
-Defining these functions with the {name}`NatProd` type name qualifying their name
+Defining these functions with the {name}`NatProd` type name qualifying their names
 allows us to use them with `.` notation:
 
 ```lean
@@ -110,7 +109,7 @@
 
 ::::full
 Since pairs will be used heavily in what follows, it will be
-convenient to write them with angle bracket notation `⟨n, m⟩`
+convenient to write them with angle-bracket notation `⟨n, m⟩`
 instead of `NatProd.pair n m`.  This notation is built into Lean and is
 called "anonymous constructor syntax".  It is available for any inductive
 type with a single constructor, as long as the expected type is declared or
@@ -443,7 +442,7 @@
 a type class is like a Java-style interface, and an _instance_ is an
 implementation of that interface for a particular type.
 We associate notation with a particular type class member, and then
-instances of that typeclass inherit the notation.
+instances of that type class inherit the notation.
 
 For example, `++` is defined via the `HAppend` type class.
 Any type that provides an {name}`HAppend` instance gets to use `++`.
@@ -816,7 +815,7 @@
 :::
 :::::
 
-Again, all these proofs could be completed with just `rfl`, because the proof is computationally straightforward — compute both sides of the equality and check if they are the same.
+Again, all these proofs could be completed with just `rfl`, because the proof is computationally straightforward — compute both sides of the equality and check whether they are the same.
 
 ```lean
 example : count 1 [1, 2, 3, 1, 4, 1] = 3 := solution!(by rfl)
@@ -1174,7 +1173,7 @@
 n :: ((l₁' ++ l₂) ++ l₃) = n :: (l₁' ++ (l₂ ++ l₃)),
 ```
 
-which is immediate from the induction hypothesis.  _Qed_.
+which is immediate from the induction hypothesis.  _QED_.
 
 ### Generalizing Statements
 
@@ -1466,13 +1465,13 @@
 ```
 
 This follows directly from the definitions of `length` and `++`
-together with the induction hypothesis.  _Qed_.
+together with the induction hypothesis.  _QED_.
 
-_Theorem_: For all lists `l`,  `l.reverse.length = l.length`.
+_Theorem_: For all lists `l`, `l.reverse.length = l.length`.
 
 _Proof_: By induction on `l`.
 
-  - First, suppose `l = []`.  We must show
+- First, suppose `l = []`.  We must show
 
 ```display
 [].reverse.length = [].length,
@@ -1506,7 +1505,7 @@
 ```
 
 This follows directly from the induction hypothesis and the
-definition of `length`.  _Qed_.
+definition of `length`.  _QED_.
 
 The style of these proofs is rather long-winded and pedantic.
 After reading a couple like this, we might find it easier to
@@ -1518,10 +1517,10 @@
 _Theorem_: For all lists `l`, `l.reverse.length = l.length`.
 
 _Proof_: First observe, by a straightforward induction on `l`,
- that `(l ++ [n]).length = .succ l.length` for any `l`.  The main
- property then follows by another induction on `l`, using the
- observation together with the induction hypothesis in the case
- where `l = n' :: l'`. _Qed_.
+that `(l ++ [n]).length = .succ l.length` for any `l`.  The main
+property then follows by another induction on `l`, using the
+observation together with the induction hypothesis in the case
+where `l = n' :: l'`. _QED_.
 
 Which style is preferable in a given situation depends on
 the sophistication of the expected audience and how similar the
@@ -1733,7 +1732,7 @@
 definition of `removeOne` above.
 ::::::
 
-::::dev "Daniel Sainati @dsainati" PotentialImprovement
+::::dev "Daniel Sainati (dsainati)" PotentialImprovement
 
 There is a nicer solution to this exercise that doesn't require the contrived
 theorem statement that has 0s instead of arbitrary numbers. The only issue
@@ -1862,7 +1861,7 @@
 
 :::::exercise (rating := 2) (name := "reverse_injective") (level := Advanced)
 Prove that {name}`reverse` is injective. Do not prove this by induction —
-that would be hard. Instead, re-use the same proof technique that
+that would be hard. Instead, reuse the same proof technique that
 you used for {name}`involutive_injective`. (But: Don't try to use that
 exercise directly as a lemma: the types are not the same!)
 
@@ -2181,7 +2180,7 @@
 Is the following claim true or false?
 
 ```lean
-theorem quiz2  (d : PartialMap) (x y : MyId) (o : Nat) :
+theorem quiz2 (d : PartialMap) (x y : MyId) (o : Nat) :
     MyId.beq x y = false →
     find x (update d y o) = find x d := by
   intro h
